pkg/uploadserver/uploadserver.go

```diff
@@ -211,6 +215,96 @@ func (app *uploadServerApp) healthzHandler(w http.ResponseWriter, r *http.Reques
 	io.WriteString(w, "OK")
 }
 
+func (app *uploadServerApp) uploadHandlerAsync(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "HEAD" {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	if r.Method != "POST" {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	if r.TLS != nil {
+		found := false
+
+		for _, cert := range r.TLS.PeerCertificates {
+			if cert.Subject.CommonName == app.clientName {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+	} else {
+		klog.V(3).Infof("Handling HTTP connection")
+	}
+
+	exit := func() bool {
+		app.mutex.Lock()
+		defer app.mutex.Unlock()
+
+		if app.uploading || app.processing {
+			w.WriteHeader(http.StatusServiceUnavailable)
+			return true
+		}
+
+		if app.done {
+			w.WriteHeader(http.StatusConflict)
+			return true
+		}
+
+		app.uploading = true
+		return false
+	}()
+
+	if exit {
+		klog.Warning("Got concurrent upload request")
+		return
+	}
+
+	cdiContentType := r.Header.Get(UploadContentTypeHeader)
+
+	klog.Infof("Content type header is %q\n", cdiContentType)
```

Comment:
Looks like this handler duplicates a lot of code with the sync version.  Any possibility to share?

Reply:
Done